\chapter{Data} % Main chapter title

\label{chapter/introduction} 
%----------------------------------------------------------------------------------------
%	SECTION 1
%----------------------------------------------------------------------------------------

\section{The Protein database}\label{sec:protein_database}
The \gls{pdb} is the central repository for experimentally determined three-dimensional structures of proteins, nucleic acids, and complexes. As of 2025, it contains over 220{,}000 entries, the majority of which are proteins solved via X-ray crystallography. Each entry includes not only the atomic coordinates of the protein structure but also extensive metadata about the experimental conditions used during crystallization.
The following outlines the data \hyperref[sec:protein_database/acquisition]{acquisition}, \hyperref[sec:protein_database/format]{format}, and \hyperref[sec:protein_database/structure]{structure} to enable data preprocessing and analysis. 

\subsection{Acquisition}\label{sec:protein_database/acquisition}
Using the RCSB Search API, all entries solved by X-ray diffraction (about 80\%) were queried. Entries that did not contain any information about the crystallization at all or the entry was otherwise incomplete were not downloaded. A Python script then converted these identifiers into download links for the mmCIF files (.cif.gz), wrote them to text files, and split the list into several parts to enable parallel downloads via wget. The complete set of structures was downloaded in compressed mmCIF format and stored locally (after decompression to .cif where needed). Due to the large number of entries, the download process took approximately two days. However, the total download time naturally depends on the available internet bandwidth.

\subsection{Format}\label{sec:protein_database/format}
The \gls{pdb} provides structural data in two formats: PDB and CIF. The newer CIF format was chosen because, unlike the fixed-width PDB format, it does not impose size limitations on large structures. In addition, it can represent complex features such as branched carbohydrates and offers greater detail and flexibility than its predecessor.
The data items are in the format of '\_' + category name + '.' + attribute name. Data categories can be either saved in either key-value or in tabular format. These can be easily parsed into a column vector as is in CSV files. For example the crystallization conditions are saved in key-value format and look like the following for 3P4V: 

\begin{verbatim}
	_exptl_crystal_grow.crystal_id      1
	_exptl_crystal_grow.method          'VAPOR DIFFUSION, SITTING DROP'
	_exptl_crystal_grow.apparatus       None
	_exptl_crystal_grow.atmosphere      None
	_exptl_crystal_grow.pH              9.5
	_exptl_crystal_grow.temp            298.0
	_exptl_crystal_grow.pdbx_details    '3.2M (NH4)2SO4, 0.1M Glycine, ...'
	_exptl_crystal_grow.time            None
\end{verbatim}

A category is stored in tabular format when a token defines multiple values. In this case, loop\_ is followed by rows of data-item names, with data values separated by whitespace. Notably, the protein’s structural information i.e., the atom\_sites category is represented in this format, as shown in the following example:

\begin{verbatim}
	loop_
	_atom_site.group_PDB
	_atom_site.id
	_atom_site.type_symbol
	_atom_site.label_atom_id
	_atom_site.label_alt_id
	_atom_site.label_comp_id
	_atom_site.label_asym_id
	_atom_site.label_entity_id
	_atom_site.label_seq_id
	_atom_site.pdbx_PDB_ins_code
	_atom_site.Cartn_x
	_atom_site.Cartn_y
	_atom_site.Cartn_z 
	ATOM   1    N  N   . VAL A 1 1   ? 6.204   16.869  4.854   1.00 49.05 ...
	ATOM   2    C  CA  . VAL A 1 1   ? 6.913   17.759  4.607   1.00 43.14 ...
	ATOM   3    C  C   . VAL A 1 1   ? 8.504   17.378  4.797   1.00 24.80 ...
	ATOM   4    O  O   . VAL A 1 1   ? 8.805   17.011  5.943   1.00 37.68 ...
\end{verbatim}

\subsection{Structure}\label{sec:protein_database/structure}

In structural biology, one rarely needs to work with all proteins deposited in the \gls{pdb}. Consequently, the \gls{pdb} is designed as an entry-oriented resource: users typically retrieve and parse the complete dataset for a single protein of interest. This design is well aligned with the standard workflow in structural biology, where researchers focus on a small number of proteins but require all available structural, experimental, and metadata associated with those entries.

In this project, however, the goal is fundamentally different: to identify patterns across many proteins, using all deposited entries in the database. Parsing a single CIF file takes \roughly 100 ms using the fastest available library, \href{https://gemmi.readthedocs.io/en/latest/index.html}{Gemmi} which amounts to approximately 9 hours for the entire PDB. Repeating such operations would be not feasible for data analysis where data typically has to be read multiple times across sessions.

To address this, the first step after downloading the database was to restructure it into a feature-based data model, in which information is grouped by category rather than by entry. Each category is stored as a separate Parquet file (also convertible to CSV), where each attribute becomes a column spanning all proteins. This enables highly targeted access: if an analysis requires only a single attribute, the corresponding column can be read directly without loading irrelevant data.

Although the initial restructuring (reading, grouping, and writing) takes \roughly 11 hours, it drastically improves downstream performance. For example, extracting all crystallization-condition information from the entire database now takes \roughly 10s, compared to the 9 hours required to parse every CIF individually.

Storage efficiency is also greatly improved. The full PDB occupies \roughly 46 GB and allocates \roughly 110 GB, making it impossible to load into memory. In contrast, the derived dataset containing only crystallization conditions is about 20 MB, small enough to be memory-resident and repeatedly queried without overhead.

This dramatic reduction in both time and storage requirements arises because most information in CIF files is irrelevant for crystallization-condition prediction. Metadata such as software provenance or publication information is not needed for this project and can be safely excluded from the feature-based representation.

However, to enable analyses that combine multiple categories, the restructured database must be joinable in a way that allows the original information to be reconstructed. This was accomplished using a star schema, in which each category table is indexed by the unique \texttt{entry\_id} of the protein. This design ensures that information from different categories (such as crystallization conditions and sequence data) can be linked by performing joins on the shared entry identifier. 
The restructured database format is depicted in \autoref{fig:proteinDatabaseStructure}.
\begin{figure}
	\centering
	\input{Media/databaseStructure}
	\caption{Restructured database format.}
	\label{fig:proteinDatabaseStructure}
\end{figure}

\subsection{Description}
After filtering for majorly incomplete entries the total number of deposited proteins was 196743 as of 01.09.2025. 
Each of these entries has around 700 attributes across 68 categories. However, as mentioned before the minority of these attributes/categories is of importance when predicting crystallization conditions. 
More specifically, the important categories and their attributes are the following: 

\subsubsection*{Input Categories}

The \texttt{entity\_poly} category of the mmCIF format describes each polymeric entity in a macromolecular structure. A polymer represents a sequence of linked monomers (e.g., amino acids or nucleotides), and this category provides essential metadata about its type, sequence, and representation in the structural model.
\texttt{entity\_id} Unique identifier linking the polymer to the corresponding entry in the \texttt{entity} category. \\
\texttt{type} Specifies the polymer type, such as polypeptide(L), DNA, RNA, or polysaccharide. \\
\texttt{nstd\_linkage} Indicates whether the polymer contains non-standard chemical linkages (e.g., cross-links or modified connectivity). \\
\texttt{nstd\_monomer} Flags the presence of non-standard or modified monomers within the sequence. \\
\texttt{pdbx\_seq\_one\_letter\_code} The polymer sequence in one-letter code, including symbols for modified residues. \\
\texttt{pdbx\_seq\_one\_letter\_code\_can} Canonicalized one-letter sequence where modified residues are mapped to their closest standard equivalents. \\
\texttt{pdbx\_strand\_id} Lists the chain identifiers (e.g., A, B) representing this polymer in the structure. \\
\texttt{pdbx\_target\_identifier} Optional external identifier used mainly in structural genomics pipelines. \\

The \texttt{atom\_site} category contains the atomic coordinates and related information that define the three-dimensional structure of the macromolecule. Each row corresponds to a single atom and records properties such as atom name, element, residue identifier, chain, Cartesian coordinates, occupancy, and atomic displacement parameters.

\subsubsection*{Label Categories}
The category \texttt{exptl\_crystal\_grow} defines the crystallization conditions. It contains 17 attributes that describe the experimental setup and methodology used to grow the protein crystal. However, as seen in \autoref{tab:missing_percentage} for 13 of the attributes the majority of entries do not contain any information. \\
 
\begin{table}[h]
	´\centering
	\begin{tabular}{lr}
		\toprule
		Attribute & Missing percentage \\
		\midrule
		method & 10.78 \\
		temp & 10.79 \\
		pH & 22.30 \\
		pdbx\_pH\_range & 75.24 \\
		pdbx\_details & 0.04 \\
		temp\_details & 98.56 \\
		apparatus & 99.84 \\
		atmosphere & 99.84 \\
		details & 99.83 \\
		method\_ref & 99.84 \\
		pressure & 99.83 \\
		pressure\_esd & 99.84 \\
		seeding & 99.83 \\
		seeding\_ref & 99.84 \\
		temp\_esd & 99.84 \\
		time & 99.84 \\
		\bottomrule
	\end{tabular}
	\caption{Percentage of Missing data per Attribute}
	\label{tab:missing_percentage}
\end{table}

Thus, the attributes of relevance are the following: \\
\texttt{method}: Describes the crystallization technique employed, such as vapor diffusion, batch crystallization, or microbatch methods (in free text).\\
\texttt{pH}: Specifies the pH of the crystallization solution, which strongly influences protein stability and crystal formation (numerical).\\
\texttt{temp}: Records the temperature at which the crystallization experiment was performed, typically given in Kelvin or Celsius (in free text).\\
\texttt{pdbx\_details}: Provides free-text experimental details, such as buffer components, precipitants, additives, or other conditions important for reproducing the crystallization setup (in free text).\\

Moreover, it might be of importance how well the protein has crystallized. Crystal quality measures are primarily provided in the \texttt{exptl\_crystal} category, which describes the physical properties of the crystal, and in the \texttt{diffrn} and \texttt{reflns} categories, which contain diffraction statistics such as resolution, completeness, and R-factors that reflect the overall quality of the crystal.

\section{Data Normalization}\label{sec:data_normalization}

\subsection{Perturbation of missing values}
Typically, 

\subsection{Details Parsing}
How the details were parsed to retrieve a uniform label representation 

\subsubsection{Pipeline}
Detailed explanation of the pipeline 

\subsubsection{Parsing Quality}
How the parsing quality was determined. 

\section{Data Analysis}\label{sec:data_analysis}

\subsection{Univariate Analysis}\label{sec:data_analysis/univariate_analysis}
Histograms, distributions, summary statistics

\subsubsection{Input}
\begin{figure}[h!]
	\centering
	
	\begin{subfigure}[b]{0.495\linewidth}
		\centering
		\includegraphics[width=\linewidth]{../msc-thesis-code/data/dataExploration/plots/sequenceLength}
		\caption{Sequence length}
		\label{fig:seqLength}
	\end{subfigure}
	\begin{subfigure}[b]{0.495\linewidth}
		\centering
		\includegraphics[width=\linewidth]{../msc-thesis-code/data/dataExploration/plots/atom_count}
		\caption{Atom Counts}
		\label{fig:atomCounts}
	\end{subfigure}
	\caption{Sequence Length and Atom Counts}
	\label{fig:dataStats}
\end{figure}


\begin{figure}[h!]
	\centering		
	\begin{subfigure}[b]{0.495\linewidth}
		\centering
		\includegraphics[width=\linewidth]{../msc-thesis-code/data/dataExploration/plots/aminoAcidOccurence}
		\caption{Amino Acid Occurence}
		\label{fig:aminoAcidOccurence}
	\end{subfigure}
	\begin{subfigure}[b]{0.495\linewidth}
		\centering
		\includegraphics[width=\linewidth]{../msc-thesis-code/data/dataExploration/plots/observedVsPredictedAminoAcidCount}
		\caption{Observed Vs. Predicted Amino Acid Count}
		\label{fig:observedVsPredictedAminoAcidCount}
	\end{subfigure}
	\caption{Sequence Length and Atom Counts}
	\label{fig:aminoAcidCounts}
\end{figure}


\begin{figure}[h!]
	\centering		
	\begin{subfigure}[b]{0.495\linewidth}
		\centering
		\includegraphics[width=\linewidth]{../msc-thesis-code/data/dataExploration/plots/gravyHist}
		\caption{Kyte-Doolittle hydropathy}
		\label{fig:gravyHist}
	\end{subfigure}
	\begin{subfigure}[b]{0.495\linewidth}
		\centering
		\includegraphics[width=\linewidth]{../msc-thesis-code/data/dataExploration/plots/aromHist}
		\caption{Aromaticity}
		\label{fig:aromaticity}
	\end{subfigure}
		\begin{subfigure}[b]{0.495\linewidth}
		\centering
		\includegraphics[width=\linewidth]{../msc-thesis-code/data/dataExploration/plots/pIHist}
		\caption{Isoelectric Point}
		\label{fig:pIHist}
	\end{subfigure}
	\begin{subfigure}[b]{0.495\linewidth}
		\centering
		\includegraphics[width=\linewidth]{../msc-thesis-code/data/dataExploration/plots/mwHist}
		\caption{Molecular Weight}
		\label{fig:mwHist}
	\end{subfigure}
	\caption{Derived Sequence Features}
	\label{fig:sequenceInformation}
\end{figure}

\subsubsection{Label}
\begin{figure}[h!]
	\centering		
	\begin{subfigure}[h]{0.495\linewidth}
		\centering
		\includegraphics[width=\linewidth]{../msc-thesis-code/data/dataExploration/plots/distribution_apolar_vs_polar}
		\caption{Apolar vs. Polar}
		\label{fig:apolarVsPolar}
	\end{subfigure}
	\begin{subfigure}[h]{0.495\linewidth}
		\centering
		\includegraphics[width=\linewidth]{../msc-thesis-code/data/dataExploration/plots/distribution_surface_information}
		\caption{Surface Distribution}
		\label{fig:surfaceDistribution}
	\end{subfigure}
		\caption{Derived Surface Features}
	\label{fig:derivedSurfaceFeatures}
\end{figure}


\subsubsection{Label}
\begin{figure}[h!]
	\centering		
	\begin{subfigure}[b]{0.495\linewidth}
		\centering
		\includegraphics[width=\linewidth]{../msc-thesis-code/data/dataExploration/plots/temperatureDistribution}
		\caption{Temperature Distribution}
		\label{fig:temperatureDistrbution}
	\end{subfigure}
	\begin{subfigure}[b]{0.495\linewidth}
		\centering
		\includegraphics[width=\linewidth]{../msc-thesis-code/data/dataExploration/plots/phDistribution}
		\caption{PH Distribution}
		\label{fig:pHDistribution}
	\end{subfigure}
	\label{fig:temperatureAndPh}
\end{figure}


\subsection{Multivariate Analysis}\label{sec:data_analysis/multivariate_analysis}
Feature-feature relationships, correlation matrices, scatter plots

\subsection{Outlier and Anomaly Detection}\label{sec:data_analysis/outlier_and_anomaly_detection}
Identifying unusual values or patterns

\subsection{Missing Data Analysis}
Assessing missingness, patterns, potential mechanism



